Sources/PostgresNIO/New/PostgresChannelHandler.swift

```diff
@@ -317,53 +323,73 @@ final class PostgresChannelHandler: ChannelDuplexHandler {
         case .sendSSLRequest:
             self.encoder.ssl()
-            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: runPromise)
         case .sendPasswordMessage(let mode, let authContext):
-            self.sendPasswordMessage(mode: mode, authContext: authContext, context: context)
+            self.sendPasswordMessage(mode: mode, authContext: authContext, context: context, promise: runPromise)
         case .sendSaslInitialResponse(let name, let initialResponse):
             self.encoder.saslInitialResponse(mechanism: name, bytes: initialResponse)
-            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: runPromise)
         case .sendSaslResponse(let bytes):
             self.encoder.saslResponse(bytes)
-            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: nil)
+            context.writeAndFlush(self.wrapOutboundOut(self.encoder.flushBuffer()), promise: runPromise)
         case .closeConnectionAndCleanup(let cleanupContext):
             self.closeConnectionAndCleanup(cleanupContext, context: context)
+            if let promise = cleanupContext.closePromise {
+                promise.futureResult.cascade(to: runPromise)
+            } else {
+                runPromise?.succeed(())
+            }
         case .fireChannelInactive:
             context.fireChannelInactive()
+            runPromise?.succeed(())
         case .sendParseDescribeSync(let name, let query, let bindingDataTypes):
-            self.sendParseDescribeAndSyncMessage(statementName: name, query: query, bindingDataTypes: bindingDataTypes, context: context)
+            self.sendParseDescribeAndSyncMessage(
+                statementName: name,
+                query: query,
+                bindingDataTypes: bindingDataTypes,
+                context: context,
+                promise: runPromise
+            )
         case .sendBindExecuteSync(let executeStatement):
-            self.sendBindExecuteAndSyncMessage(executeStatement: executeStatement, context: context)
+            self.sendBindExecuteAndSyncMessage(executeStatement: executeStatement, context: context, promise: runPromise)
         case .sendParseDescribeBindExecuteSync(let query):
-            self.sendParseDescribeBindExecuteAndSyncMessage(query: query, context: context)
+            self.sendParseDescribeBindExecuteAndSyncMessage(query: query, context: context, promise: runPromise)
         case .succeedQuery(let promise, with: let result):
-            self.succeedQuery(promise, result: result, context: context)
+            self.succeedQuery(rowsPromise: promise, result: result, context: context, promise: runPromise)
```

Comment:
`succeedQuery()` currently doesn't do much complicated and I could just succeed `runPromise` here, but I decided to pass `runPromise` to it for implying that the function is responsible of it.